instrumentctl/g9_driver.py

```diff
@@ -0,0 +1,264 @@
+inStatus = {
+    0: "No error",
+    1: "Invalid configuration",
+    2: 'External test signal failure',
+    3: 'Internal circuit error',
+    4: 'Discrepancy error',
+    5: 'Failure of the associated dual-channel input'
+}
+
+outStatus = {
+    0: 'No error',
+    1: 'Invalid configuration',
+    2: 'Overcurrent detection',
+    3: 'Short circuit detection',
+    4: 'Stuck-at-high detection',
+    5: 'Failure of the associated dual-channel output',
+    6: 'Internal circuit error',
+    8: 'Dual channel violation'
+}
+
+usStatus = {
+    9: "Output Power Supply Error Flag",
+    10: "Safety I/O Terminal Error Flag",
+    13: "Function Block Error Flag"
+}
+
+class G9Driver:
+    def __init__(self, port=None, baudrate=9600, timeout=0.5, logger=None, debug_mode=False):
+        if port:
+            self.ser = serial.Serial(port, baudrate, parity=serial.PARITY_EVEN, stopbits=serial.STOPBITS_ONE, bytesize=serial.EIGHTBITS, timeout=timeout)
+        self.debug_mode = debug_mode
+        self.logger = logger
+        self.lastResponse = None
+        self.msgOptData = None
+        self.input_flags = None
+
+    def send_command(self):
+        if not self.is_connected():
+            raise ConnectionError("Seiral Port is Not Open.")
+        header = b'\x40\x00\x00\x0F\x4B\x03\x4D\x00\x01' 
+        data = b'\x00\x00\x00\x00'
+        reserve = b'\x00\x00'
+        self.msgOptData = data
+        checksum = self.calculate_checksum(header + data + reserve, 0 , len(header + data))
+        cs = b'\x00' + checksum if len(checksum) == 1 else checksum
+        footer = b'\x2A\x0D' 
+        #TODO: add try and catch
+        self.ser.write(header + data + reserve+ cs + footer)
+
+        self.response()
+
+    # used mainly for the check sum but can also be used to check for error flags
+    # needs an input of a byte string and the range of bytes that need to be sum
+    # will return the sum of the bytes in the a byte string in the form of b'\x12'
+    def calculate_checksum(self, byteString, startByte, endByte):
+        assert isinstance(byteString, bytes)
+        return sum(byteString[startByte:endByte + 1]).to_bytes(1, "big") 
+
+    # helper function to convert bytes to bits for checking flags
+    # not currently being used but many be helpful in the future for getting errors
+    def bytes_to_binary(self, byte_string):
+        return ''.join(format(byte, '08b') for byte in byte_string)
+    
+    # this method is made to check the error flags, right not only checks the last 13 bits
+    # of a byte string
+    def check_flags13(self, byteString, norm = 1):
+        assert isinstance(byteString, bytes)
+        # this is for if we only need the last 13 bits (more or less hardcoding this 
+        # just including the rest if it might be helpful in the future
+        if int(self.bytes_to_binary(byteString)[-13:], 2) >= (13 * norm):
+            # all flags we care about are 1
+            return True
+        else:
+            return False
```

Comment:
My understanding is that this method might be used to check if the status bits in the given bytes object are all set to 1. but it looks like there's a type conversion problem here that might cause problems.  

The condition `if int(bits[-13:], 2) >= (13 * norm):` only checks if the integer value of the bitstring >= 13, not that all the status flags are 1's. The check will incorrectly return True on some bitstrings that contain 0's (error flags)

converting 13 bits into an integer you get 2^13 - 1= 8191
```
>>>
>>> def check_flags13(byteString, norm=1):
...     assert isinstance(byteString, bytes)
...     binary_string = bytes_to_binary(byteString)
...     last_13_bits = binary_string[-13:]
...     integer_value = int(last_13_bits, 2)
...     if integer_value >= (13 * norm):
...         return True
...     else:
...         return False
...
>>> def bytes_to_binary(byte_string):
...     return ''.join(format(byte, '08b') for byte in byte_string)
...
>>> byteString = bytes([0xFF, 0xFF])
>>> print(bytes_to_binary(byteString))
1111111111111111
>>> check_flags13(byteString)
True
>>> errorByteString = bytes([0x00, 0x1F])
>>> print(bytes_to_binary(errorByteString))
0000000000011111
>>> check_flags13(errorByteString)
True
>>>
```

Reply:
I agree it seems like checking >=13 is not the same as all flags being 1

Reply:
We're comparing the string value of the binarystring to that of 13 1's now.

Reply:
In the int casing this coverts the int from binary to decimal with the 2 argument. There was two reasons I went with this method, one being that the flags are not the same for all inputs for some 1 means error and others 0 means error. Thus this method could be used for both given the norm argument. Along with that it also gives an easier way to change the number of inputs. But if we think a different way is better, we can go with that.

![Screenshot 2024-10-17 081223](https://github.com/user-attachments/assets/8ef4dc1f-42c9-4c21-a34c-6f4e7e02862a)